server/src/main/java/org/opensearch/indices/replication/SegmentReplicationSourceService.java

```diff
@@ -86,6 +105,25 @@ public SegmentReplicationSourceService(
     private class CheckpointInfoRequestHandler implements TransportRequestHandler<CheckpointInfoRequest> {
         @Override
         public void messageReceived(CheckpointInfoRequest request, TransportChannel channel, Task task) throws Exception {
+            ShardId shardId = request.getCheckpoint().getShardId();
+            final IndexService indexService = indicesService.indexService(shardId.getIndex());
+            final IndexShard indexShard = indexService.getShard(shardId.id());
+
+            // If not already tracked, initiate tracking of the new replica.
+            // At this point the replica will start receiving requests.
+            if (indexShard.getReplicationGroup().getTrackedAllocationIds().contains(request.getTargetAllocationId()) == false) {
```

Comment:
Are there other use-cases, which initiates replica tracking ? If yes, in that case, can we get rid of this handling

Reply:
Nope only this class and peer recovery's `RecoverySourceHandler`.  `CheckpointInfoRequestHandler` is invoked on every replication event.  And every event for a replica other than the first, the allocation id will already be tracked.  This is more importantly to avoid refreshing the primary in these instances on line 125.  We only need this extra refresh when recovering a new shard, and this check tells us if that has already happened.